File: src/main/scala/com/scala/leetcode/array/test/LongestCommonPrefix.scala
package com.scala.leetcode.array.test

object LongestCommonPrefix {

  def main(args: Array[String]): Unit = {
    println("LongestCommonPrefix is : " + longestCommonPrefix(Array("flower", "flow", "flight")))
    println("LongestCommonPrefix is : " + longestCommonPrefix(Array("dog", "racecar", "car")))
    println("LongestCommonPrefix is : " + longestCommonPrefix(Array("a")))
    println("LongestCommonPrefix is : " + longestCommonPrefix(Array("flower","flower","flower","flower")))
    println("LongestCommonPrefix is : " + longestCommonPrefix(Array("aac","aa","ccc")))
  }

  def findCommonPrfixNew(arr: Array[String]): String = {

    var ret = ""
    var charLen = 0
    var endLen = 0
    
    if(arr.toSet.size == 1) return arr { 0 }
    
    
//        for (i <- 1 until arr.length) {
//          println(arr { 0 } +", " +arr{i})
//        }
      
    
    try {
      for (i <- 1 until arr.length) {
        println(arr { 0 }.charAt(charLen) + ", " + arr { i }.charAt(charLen))
        if (arr { 0 }.charAt(charLen) == arr { i }.charAt(charLen)) {
          charLen = charLen + 1
        }
      }
      

     if (endLen < charLen) {
        endLen = charLen
      }
      
      charLen = 0
      
      println("charLen " + charLen + ", endLen " + endLen)
    } catch {
      case e: StringIndexOutOfBoundsException => 
        println("StringIndexOutOfBoundsException")
        return ret
    }
    arr { 0 }.substring(0, endLen)
  }

  def findCommonPrfix(arr: Array[String]): String = {

    var ret = ""
    var aSet = arr
    try {
      for (i <- 0 until aSet.length; j <- 1 until aSet.length) {
        //println(aSet { i })
        for (in <- 0 until 3) {
          if (aSet { i }.charAt(in) == aSet { j }.charAt(in)) {
            // println(aSet { i }.charAt(in))
            ret = ret + aSet { i }.charAt(in)
          }
        }
      }
    } catch {
      case e: StringIndexOutOfBoundsException => return ret
    }
    ret
  }
  
  def longestCommonPrefix(strs : Array[String]) : String ={
    
    if(strs == null || strs.length ==0)  return ""
    
    for(i<- 0 until strs{0}.length(); j<-1 until strs.length){
      var c = strs{0}.charAt(i)
      if (i == strs{j}.length() || strs{j}.charAt(i) != c)
                return strs{0}.substring(0, i); 
    }
    return strs{0}
  }

}

File: src/main/java/com/leetcode/string/LongestCommonPrefix.java
package com.leetcode.string;

import java.util.Arrays;

/**
 * Write a function to find the longest common prefix string amongst an array of strings.
 * If there is no common prefix, return an empty string "".
 *
 * Example 1:
 *
 * Input: strs = ["flower","flow","flight"]
 * Output: "fl"
 *
 * Example 2:
 * Input: strs = ["dog","racecar","car"]
 * Output: ""
 * Explanation: There is no common prefix among the input strings.
 *
 * Constraints:
 * 1 <= strs.length <= 200
 * 0 <= strs[i].length <= 200
 * strs[i] consists of only lowercase English letters.
 */
public class LongestCommonPrefix {
    public static void main(String[] args) {
        System.out.println("Longest common prefix: " + longestCommonPrefix(new String[]{"flower", "flow", "flight"}));
        System.out.println("Longest common prefix with sorting: " + longestCommonPrefixWithSorting(new String[]{"flower", "flow", "flight"}));

    }

    public static String longestCommonPrefix(String[] strs) {
        if (strs == null || strs.length == 0) {
            return "";
        }
        String prefix = strs[0];
        for (int i = 0; i < prefix.length(); i++) {
            char c = prefix.charAt(i);
            for (int j = 1; j < strs.length; j++) {
                if (i == strs[j].length() || c != strs[j].charAt(i)) {
                    return prefix.substring(0, i);
                }

            }

        }
        return prefix;
    }

    public static String longestCommonPrefixWithSorting(String[] strs) {
        Arrays.sort(strs, (a,b) -> Integer.compare(a.length(), b.length()));
        for (int i = 0; i < strs[0].length(); i++) {
            if(i == strs[1].length() || strs[0].charAt(i) != strs[1].charAt(i)) {
                return strs[0].substring(0, i);
            }
        }
        return strs[0];
    }
}
